libpromises/evalfunction.c

```diff
@@ -8336,6 +8336,128 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+    if (copied >= sizeof(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, path is too long (%zu > %zu)",
+            fp->name, copied, sizeof(path));
+        return FnFailure();
+    }
+
+    if (!IsAbsoluteFileName(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, not an absolute path '%s'",
+            fp->name, path);
+        return FnFailure();
+    }
+
+    if (!IsDir(path))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, path '%s' is not a directory",
+            fp->name, path);
+        return FnFailure();
+    }
+
+    DeleteRedundantSlashes(path);
+
+    int len = strlen(path);
+    while (path[len - 1] == FILE_SEPARATOR && len > 0)
+    {
+        path[--len] = '\0';
+    }
+
+    if (finalargs->next == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s require glob as second argument",
+            fp->name);
+        return FnFailure();
+    }
+
+    char glob[PATH_MAX];
+    copied = strlcpy(glob, RlistScalarValue(finalargs->next), sizeof(glob));
+    if (copied >= sizeof(glob))
+    {
+        Log(LOG_LEVEL_ERR, "Function %s, glob is too long (%zu > %zu)",
+            fp->name, copied, sizeof(glob));
+        return FnFailure();
+    }
+
+    if (IsAbsoluteFileName(glob))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Function %s, glob '%s' cannot be an aboslute path", fp->name,
+            glob);
+        return FnFailure();
+    }
+
+    DeleteRedundantSlashes(glob);
+
+    /* level defaults to inf */
+    long level = IntFromString("inf");
+    if (finalargs->next->next != NULL)
+    {
+        level = IntFromString(RlistScalarValue(finalargs->next->next));
+    }
+
+    JsonElement *json = JsonArrayCreate(1);
+
+    while (level-- >= 0)
+    {
+        char filepath[PATH_MAX];
+        int ret = snprintf(filepath, sizeof(filepath), "%s%c%s", path,
+                           FILE_SEPARATOR, glob);
+        if (ret < 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Function %s, unexpected error occured while formatting "
+                "filepath: %s", fp->name, GetErrorStr());
+            return FnFailure();
+        }
+        else if ((size_t) ret >= sizeof(filepath))
+        {
+            Log(LOG_LEVEL_ERR,
+                "Function %s, truncation error occured while formatting "
+                "filepath", fp->name);
+            return FnFailure();
+        }
+
+        StringSet *files = GlobFileList(filepath);
+        StringSetIterator iter = StringSetIteratorInit(files);
+        const char *file = StringSetIteratorNext(&iter);
+        while(file != NULL)
+        {
+            JsonArrayAppendString(json, file);
+            file = StringSetIteratorNext(&iter);
+        }
+        StringSetDestroy(files);
+
+        char *sep = strrchr(path, FILE_SEPARATOR);
+        if (sep == NULL)
+        {
+            /* don't search beyond root directory */
+            break;
+        }
+        *sep = '\0';
+    }
+
+    return FnReturnContainerNoCopy(json);
```

Comment:
Please add a note about why this function returns `data/JSON` to the commit message. It's not quite straightforward.